[GB][E2E] Introduce a simpler upgrade smoke test E2E for blocks in simple and AT sites

fullofcaffeine opened the pull request:
#### Changes proposed in this Pull Request

Project: p9oQ9f-13V-p2.

Based on the original discussion with @dpasque: p1643899966346489-slack-C01QG4Y91RR.

* Replaces the [old, more complex Gutenberg upgrade E2E spec](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/wpcom-old/gutenberg__upgrade.js) with a simpler Playwright-based version.
* The new spec relies on a pre-saved Post in a specific edge site that already has all the most-popular blocks configured.
* We focus on verifying if there are any warnings rendered in the editor (which means an invalidation or actual block crash) and fail if we find at least one.
* We don't check the frontend anymore. Checking the frontend requires more data about the block and is best left out for specific block tests _(preferably upstream!)_. We are testing that the blocks continue to work across GB upgrades and verifying if they crash in the editor only is enough. _If we really want to verify the rendering in the frontend, it might be better to follow up with a vizreg kind of test for that (TBD)._
* We don't rely on two sites _(regular,edge)_ anymore. Instead, we rely on a single edge site for simple and another for AT. These sites will get updated weekly with new GB versions.

### Build runs

This test will be part of the `Playwright Gutenberg E2E tests` builds. These builds are triggered everytime we update simple AT with the edge GB version. These builds also run daily, so we should be able to catch AT failures caused by other components (i.e JetPack, wpcomsh, ETK, Calypso) before we run them again in the next edge update.

Related diff: D75693-code

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Tests should pass.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

fullofcaffeine committed:
Add new, simpler, Playwright-based E2E spec to smoke test blocks across GB versions (current->edge)

fullofcaffeine committed:
Skip test if GUTENBERG_EDGE not set

fullofcaffeine committed:
Test on AT and simple

fullofcaffeine committed:
Cleanup

fullofcaffeine committed:
Improve test description

fullofcaffeine committed:
Refactor the GutenbergEditorPage.getEditorFrame() function to support falling back to the page's main frame if the iframe can't be found

fullofcaffeine commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -137,22 +137,34 @@ export class GutenbergEditorPage {
+				// the Gutenframe, so we return the `mainFrame` to allow the test to access
+				// the editor in the WPAdmin page.
+				console.info( 'Could not locate editor iframe. Returning the top-level frame instead' );
+				return await this.page.mainFrame();
```
What led me to try this was the fact that my AT test site was redirecting to WPadmin after a few seconds. The test would then fail after the timeout because the Gutenframe couldn't be found. This works well as a fallback for when Calypso ends up redirecting to WP-Admin. It will return the top-level frame where the Gutenberg editor can be found and everything else will work accordingly.

I'm not 100% sure about this yet, though. I can see a few potential issues:
1. One is of test intent - one could argue that abstracting this logic is not good because we should know when the iframe is not loading and the test should fail. However, you could also argue that this component is interested in the editor and not how to get hold of it, and the actual testing of the iframe logic should be done in a separate test/test component;
2. Is of creating a potential bottleneck. The fallback only takes place after 105000ms (when the timeout exceptions are finally thrown). That can add up to the test time. Could we possibly reduce the timeout?

It might as well be better to assume that simple and AT (test) sites will always load through Calypso (and hence, the Gutenframe) and if it's redirected to WPAdmin, then something is wrong. Then, we could have a specific suite of tests that test Gutenframe itself (though it might be better to look into adding it as further down the test pyramid, at first glance, I can't find any tests for the https://github.com/Automattic/wp-calypso/blob/trunk/client/gutenberg/editor/calypsoify-iframe.tsx nor https://github.com/Automattic/wp-calypso/blob/trunk/apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js. In the end, it might just be overkill to work on this now, though it's worth discussing.

@Automattic/quality-squad looking forward to reading your thoughts about this.

dpasque replied:
Oh man, I waffle back and forth on this a lot! And have for this specific question of getting booted out of the Gutenframe.

I think my current take is that having options to add resiliency to tests that involve the editor is probably the bigger win. We struggle more with false positives than false negatives currently. So I would favor adding support to have the tests still work if we've been booted from the Gutenframe.

You do bring up the good point though that we probably want the ability to write tests that catch when the Gutenframe isn't loading. 

Can we have the best of both worlds? Could we expose an option in the constructor of the `GutenbergEditorPage` that controls whether this throws or not? That way, tests that need the resiliency could have it, but we also have an easy way to build out specs that specifically look for the redirect away from the Gutenframe. This feels like an appropriate use of constructor options to me personally!

fullofcaffeine replied:
@dpasque Thanks for the reply! 

> Can we have the best of both worlds? Could we expose an option in the constructor of the GutenbergEditorPage that controls whether this throws or not? That way, tests that need the resiliency could have it, but we also have an easy way to build out specs that specifically look for the redirect away from the Gutenframe. This feels like an appropriate use of constructor options to me personally!

Sounds good! From what I understood, we'd have this resiliency logic activated by default, right?

fullofcaffeine replied:
@dpasque Done 79738e23f5aff43c0835012d9f068a4431bc9b46. Let me know what you think :)

dpasque replied:
>we'd have this resiliency logic activated by default, right?

Yes, that's what I imagined!

>@dpasque Done https://github.com/Automattic/wp-calypso/commit/79738e23f5aff43c0835012d9f068a4431bc9b46. Let me know what you think :)

I think this looks really nice!! 🎉  Thanks for adding that! 😄 

My one suggestion is maybe we could call it `throwIfNotGutenframe` because there may be cases where the `/wp-admin` route is loaded directly? As in, someone may not actually get the boot! 🥾

fullofcaffeine replied:
> My one suggestion is maybe we could call it throwIfNotGutenframe because there may be cases where the /wp-admin > route is loaded directly? As in, someone may not actually get the boot! hiking_boot

Ah, that makes sense!

fullofcaffeine replied:
I followed your suggestion in principle but called the param/property `requiresGutenframe`. I personally don't like naming things with a negation, as it makes it can make actually negating them in the code a bit confusing to reason about. I've done it in https://github.com/Automattic/wp-calypso/pull/61349/commits/a688ea19057334b5e4878efff668d87375280cd9 (check the commit description also). Let me know your thoughts about it!

dpasque replied:
Ooo, that reads so nicely!! Love it! 🎉

fullofcaffeine committed:
Remove the old GB upgrade test

fullofcaffeine committed:
Make the Gutenframe resiliency logic the default and allow to deeactivate it through contructor options in `GutenbergEditorPage`

fullofcaffeine committed:
Rename options param to `requiresGutenframe`

The previous name suggested the error would only happen if the user was booted from the Gutenframe, but it might be the case that a direct navigation to wp-admin is made (i.e if the site is configured to not use the Gutenframe), in this case we might still want the test to "require" the Gutenframe if we're testing an aspect related to it, and fail if the Gutenframe is not present. This name better reflects those semantics.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -137,22 +155,40 @@ export class GutenbergEditorPage {
+			.elementHandle( {
+				timeout: 105 * 1000,
+			} )
+			.then( async ( elementHandle ) => ( await elementHandle!.contentFrame() ) as Frame )
+			.catch( async () => {
+				if ( this.requiresGutenframe ) {
+					throw new Error( 'Could not locate editor iframe' );
+				} else {
+					// The CalipsoifyIframe component will redirect to the plain WPAdmin page if
+					// there's an error or if the iframe takes too long to completely load. The
+					// goal here is to ge hold of the editor and not test any aspects related to
+					// the Gutenframe, so we return the `mainFrame` to allow the test to access
+					// the editor in the WPAdmin page.
+					console.info( 'Could not locate editor iframe. Returning the top-level frame instead' );
+					return await this.page.mainFrame();
+				}
+			} );
```
Tiny nit, but in our style guide we've been trying to avoid promise chaining and instead using the `async/await` syntax. Would you be up for updating that here?

fullofcaffeine replied:
Sure!

dpasque commented on test/e2e/specs/blocks/blocks__smoke-test-across-gb-edge-versions.ts:
```diff
@@ -0,0 +1,58 @@
+import { GutenbergEditorPage, TestAccount, envVariables } from '@automattic/calypso-e2e';
+import { Page, Browser } from 'playwright';
+
+const test = envVariables.GUTENBERG_EDGE ? it : it.skip;
```
I love this kind of syntax, but we actually already have a couple wrappers for this already [here](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/jest-conditionals.ts)!

Would you be up for using one of those two just for similarity among specs?

fullofcaffeine replied:
Sure!

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -137,22 +155,40 @@ export class GutenbergEditorPage {
+		if ( ! this.requiresGutenframe ) {
+			// If we're not in the Gutenframe context (i.e not Calypso), then
+			// just return the top frame, no need to try to locate the iframe.
+			if ( ! this.page.url().startsWith( 'https://wordpress.com' ) ) {
```
Ack, it looks like this won't work on PR builds because of the Calypso live URL. This is why the PR e2e builds failed on this PR.

Hows about searching for `wp-admin` in the URL instead?

fullofcaffeine replied:
Ah, good point! I'll do that 👍🏻

dpasque reviewed (commented):
🎉  This is beautiful!! So simple and elegant! Thank you for doing this! 🙌 

Got wwo small nits, and a fix for the PR failures.

fullofcaffeine committed:
Write logic using async/await instead of using Promise chaining

fullofcaffeine committed:
Look for "/wp-admin" instead of the WPCOM URL to check if running for wp-admin as the WPCOM url will not work in PRs (due to the calypso.live URLs)

fullofcaffeine committed:
Use the "skipItIf" helper

fullofcaffeine committed:
Merge branch 'trunk' into refactor/gb-upgrade-test-pw-and-simplify

fullofcaffeine committed:
Merge branch 'trunk' into refactor/gb-upgrade-test-pw-and-simplify

dpasque reviewed (approved):
Looks awesome!! Thanks for doing this! 🎉
